### PR TITLE
tracer support for contract interoperability

### PR DESCRIPTION
**Pull request by saratnt**
## Description
Added some tests related to collecting execution traces when there are calls from EVM to native smart contracts and vice-versa.

## Jira Ticket
[SDK-1352](https://horizenlabs.atlassian.net/browse/SDK-1352)

## Changes

- Added unit tests 
- Updated python tests for calls from native smart contracts to EVM smart contracts

## Checks
- [ X] Project Builds
- [ X] Project passes tests and checks
- [ ] Updated documentation accordingly
- [ ] Breaking changes have been correctly tagged and notified
 
## Additional information
It needs the following pull requests:
- libevm https://github.com/HorizenOfficial/libevm/pull/18
- geth fork https://github.com/HorizenOfficial/go-ethereum/pull/2

[SDK-1352]: https://horizenlabs.atlassian.net/browse/SDK-1352?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by alsala — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/io/horizen/account/state/ProxyMsgProcessor.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  def doInvokeSmartContractCallCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext): Array[Byte] = {
+    doInvokeSmartContractCmd(invocation, view, context, readOnly = false)
+  }
+
+  private def doInvokeSmartContractCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext, readOnly : Boolean): Array[Byte] = {
+
+    val msg = context.msg
+    log.debug(s"Entering with invocation: $invocation")
+    // check that message contains a nonce, in the context of RPC calls the nonce might be missing
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/io/horizen/account/state/ProxyMsgProcessor.scala:39]: Done

File: sdk/src/main/scala/io/horizen/account/state/ProxyMsgProcessor.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    val contractAddress = cmdInput.contractAddress
+    val data = cmdInput.dataStr
+
+    if (view.isEoaAccount(contractAddress)) {
+      throw new ExecutionRevertedException(s"smart contract address is an EOA")
+    }
+
+    val dataBytes = Numeric.hexStringToByteArray(data)
+    val additionalDepth = 0
+    val res = context.executeDepth(
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/io/horizen/account/state/ProxyMsgProcessor.scala:68]: Done

File: sdk/src/main/scala/io/horizen/account/state/Invocation.scala [hunk 3]
```diff
@@ -55,10 +57,23 @@ case class Invocation(
     } else if (readOnly) {
       TracerOpCode.STATICCALL
     } else {
-      TracerOpCode.CREATE
+      TracerOpCode.CALL
     }
   }
 
+  override def toString: String =
```
Comment (davehorizen) [inline sdk/src/main/scala/io/horizen/account/state/Invocation.scala:64]: Is this method currently used in the SDK implementation?

File: qa/sc_evm_native_interop.py [hunk 4]
```diff
@@ -157,6 +174,163 @@ def run_test(self):
...
+                "to": native_contract_address,
+                "input": data_input
+            }, "latest"
+        )
+
+        value = int(expected_value['result'], 16)
+        assert_equal(initial_value, value)
+
+        # Verify tracing
+        trace_response = node.rpc_debug_traceCall(
```
Comment (davehorizen) [inline qa/sc_evm_native_interop.py:225]: I think we need to test the tracer call also with the other built-in tracers, at least the opcode logger and the 4byteTracer just to check that everything works correctly.

File: qa/sc_evm_native_interop.py [hunk 5]
```diff
@@ -157,6 +174,163 @@ def run_test(self):
...
+            {
+                "to": native_contract_address,
+                "input": data_input_failed
+            }, "latest"
+        )
+
+        assert_true('error' in result)
+        assert_true('write protection' in result['error']['message'])
+
+        # TODO traceCall seems to have problems with failed txs. Needs to check what is the correct behavior
```
Comment (davehorizen) [inline qa/sc_evm_native_interop.py:303]: Is this expected to be covered with this branch?

**Review by davehorizen — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/io/horizen/account/state/Invocation.scala (hunk 3 above)
Comment (saratnt) [reply sdk/src/main/scala/io/horizen/account/state/Invocation.scala:64]: I don't know but it is possible in some logging.

File: qa/sc_evm_native_interop.py (hunk 5 above)
Comment (saratnt) [reply qa/sc_evm_native_interop.py:303]: No, there is a Jira ticket for this issue: https://horizenlabs.atlassian.net/browse/SDK-1434
It is solved in this PR:
https://github.com/HorizenOfficial/Sidechains-SDK/pull/924

File: qa/sc_evm_native_interop.py (hunk 4 above)
Comment (saratnt) [reply qa/sc_evm_native_interop.py:256]: Good idea. Done.

**Review by alsala — APPROVED** [review]

**Pull request closed without merging**